crates/pypi-types/src/requirement.rs

```diff
@@ -415,3 +432,229 @@ impl RequirementSource {
+    type Error = RequirementError;
+
+    fn try_from(wire: RequirementSourceWire) -> Result<RequirementSource, RequirementError> {
+        match wire {
+            RequirementSourceWire::Registry { specifier, index } => {
+                Ok(RequirementSource::Registry { specifier, index })
+            }
+            RequirementSourceWire::Git { git } => {
+                let mut url = Url::parse(&git)?;
+
+                let mut reference = GitReference::DefaultBranch;
+                let mut subdirectory = None;
+                for (key, val) in url.query_pairs() {
+                    match &*key {
+                        "tag" => reference = GitReference::Tag(val.into_owned()),
+                        "branch" => reference = GitReference::Branch(val.into_owned()),
+                        "rev" => reference = GitReference::from_rev(val.into_owned()),
+                        "subdirectory" => subdirectory = Some(PathBuf::from(val.into_owned())),
+                        _ => continue,
+                    };
+                }
+
+                let precise = url.fragment().map(GitSha::from_str).transpose()?;
+
+                url.set_query(None);
+                url.set_fragment(None);
+
+                Ok(RequirementSource::Git {
+                    repository: url.clone(),
+                    reference,
+                    precise,
+                    subdirectory,
+                    url: VerbatimUrl::from_url(url),
+                })
+            }
+            RequirementSourceWire::Direct { url, subdirectory } => Ok(RequirementSource::Url {
+                url: VerbatimUrl::from_url(url.clone()),
+                subdirectory: subdirectory.map(PathBuf::from),
+                location: url.clone(),
+            }),
+            RequirementSourceWire::Path { path } => {
+                let path = PathBuf::from(path);
+                Ok(RequirementSource::Path {
+                    url: VerbatimUrl::from_path(path.as_path())?,
+                    install_path: path.clone(),
+                    lock_path: path,
+                })
+            }
+            RequirementSourceWire::Directory { directory } => {
+                let directory = PathBuf::from(directory);
+                Ok(RequirementSource::Directory {
+                    url: VerbatimUrl::from_path(directory.as_path())?,
+                    install_path: directory.clone(),
+                    lock_path: directory,
+                    editable: false,
+                })
+            }
+            RequirementSourceWire::Editable { editable } => {
+                let editable = PathBuf::from(editable);
+                Ok(RequirementSource::Directory {
+                    url: VerbatimUrl::from_path(editable.as_path())?,
+                    install_path: editable.clone(),
+                    lock_path: editable,
+                    editable: true,
+                })
+            }
+        }
+    }
+}
```

Comment:
If we're using a Git URL here, perhaps we should in the lock file too? Then I could just close https://github.com/astral-sh/uv/pull/4631.

Reply:
That seems ok to me.

Reply:
Let me re-review that quickly, I think I lost it.

Reply:
Honestly, I'm fine with either!